File: FreeNow/Modules/VehicleList/View/VehicleListViewController.m
//
//  VehicleListViewController.m
//  FreeNow
//
//  Created by Varol Aksoy on 1.02.2022.
//

#import "FreeNow-Swift.h"
#import "VehicleListViewController.h"

@interface VehicleListViewController ()<VehicleListViewModelDelegate>
@end

@implementation VehicleListViewController
static NSString *cellIdentifier = @"VehicleListCollectionViewCell";

- (void)viewDidLoad {
    [super viewDidLoad];
    _viewModel.delegate = self;
    [_viewModel load];
}

// MARK: - UICollectionViewDataSource

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return [self.viewModel numberOfItems];
}
 
- (__kindof UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    VehicleListCollectionViewCell *cell =
    [self.vehicleListCollectionView dequeueReusableCellWithReuseIdentifier:cellIdentifier forIndexPath:indexPath];
    
    Vehicle *vehicle = [self.viewModel vehicles:(indexPath.item)];
    VehicleListCollectionViewCellViewModel *cellVM = [[VehicleListCollectionViewCellViewModel alloc] initWithVehicle:vehicle];

    cell.viewModel = cellVM;
    return cell;
}
 
// MARK: - UICollectionViewDelegate

- (void)prepareCollectionView{
    self.vehicleListCollectionView.backgroundColor = [UIColor whiteColor];
    self.vehicleListCollectionView.delegate = self;
    self.vehicleListCollectionView.dataSource = self;
    [self.vehicleListCollectionView registerNib:[UINib nibWithNibName:cellIdentifier bundle:nil] forCellWithReuseIdentifier:cellIdentifier];
}

- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath {
    CGFloat fullWidth = CGRectGetWidth(collectionView.bounds);
    return CGSizeMake(fullWidth, [self.viewModel heightDimension]);
}

// MARK: - VehicleListViewModelDelegate

- (void)showLoadingView {
    [[LoadingView shared] startLoading];
}

- (void)hideLoadingView {
    [[LoadingView shared] hideLoading];
}

- (void)reloadData {
    [self.vehicleListCollectionView reloadData];
}

- (void)addRefreshControl {
    UIRefreshControl *refreshControl = [[UIRefreshControl alloc] init];
    refreshControl.tintColor = [UIColor grayColor];
    [refreshControl addTarget:self action:@selector(pullToRefresh:) forControlEvents:UIControlEventValueChanged];
    [self.vehicleListCollectionView addSubview:refreshControl];
    self.vehicleListCollectionView.alwaysBounceVertical = YES;

}

- (void) pullToRefresh:(id)sender {
    [self.viewModel pullToRefresh];
    
    UIRefreshControl *refreshControl = (UIRefreshControl *)sender;
    if (refreshControl) {
        [refreshControl endRefreshing];
    }
}

@end

File: FreeNow/Modules/VehicleList/View/VehicleListViewController.swift
//
//  VehicleListViewController.swift
//  FreeNow
//
//  Created by Varol Aksoy on 28.01.2022.
//

import UIKit

final class VehicleListViewController: BaseViewController, LoadingShowable {
    @IBOutlet private weak var vehicleListCollectionView: UICollectionView!

    var viewModel: VehicleListViewModel! {
        didSet {
            viewModel.delegate = self
        }
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        viewModel.load()
    }
    
    @objc fileprivate func pullToRefresh() {
        viewModel.pullToRefresh()
    }
}

extension VehicleListViewController: UICollectionViewDataSource {
    func collectionView(_ collectionView: UICollectionView,
                        numberOfItemsInSection section: Int) -> Int {
        
        viewModel.numberOfItems
    }
    
    func collectionView(_ collectionView: UICollectionView,
                        cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        
        let cell = collectionView.dequeCell(cellType: VehicleListCollectionViewCell.self,
                                            indexPath: indexPath)
        
        if let vehicle = viewModel.vehicles(indexPath.item) {
            cell.viewModel = VehicleListCollectionViewCellViewModel(vehicle: vehicle)
        }
        
        return cell
    }
}

extension VehicleListViewController: UICollectionViewDelegate {
    override func viewWillTransition(to size: CGSize,
                                     with coordinator: UIViewControllerTransitionCoordinator) {
        super.viewWillTransition(to: size, with: coordinator)
        coordinator.animate(alongsideTransition: { context in
            self.vehicleListCollectionView.collectionViewLayout.invalidateLayout()
        }, completion: nil)
    }
}

extension VehicleListViewController: VehicleListViewModelDelegate {
    func showLoadingView() {
        showLoading()
    }
    
    func hideLoadingView() {
        hideLoading()
    }
    
    func reloadData() {
        vehicleListCollectionView.reloadData()
    }
    
    func endRefreshing() {
        vehicleListCollectionView.refreshControl?.endRefreshing()
    }
    
    func addRefreshControl() {
        let refreshControl = UIRefreshControl()
        refreshControl.addTarget(self, action: #selector(pullToRefresh), for: .valueChanged)
        vehicleListCollectionView.refreshControl = refreshControl
    }
    
    func prepareCollectionView() {
        vehicleListCollectionView.dataSource = self
        vehicleListCollectionView.delegate = self
        vehicleListCollectionView.register(cellType: VehicleListCollectionViewCell.self)
        
        handleCompositionalLayout()
    }
    
    fileprivate func handleCompositionalLayout() {
        let size = NSCollectionLayoutSize(
            widthDimension: NSCollectionLayoutDimension.fractionalWidth(viewModel.widthDimension),
            heightDimension: NSCollectionLayoutDimension.estimated(viewModel.heightDimension)
        )
        
        let item = NSCollectionLayoutItem(layoutSize: size)
        let group = NSCollectionLayoutGroup.horizontal(layoutSize: size,
                                                       subitem: item,
                                                       count: viewModel.layoutGroupCount)
        
        let section = NSCollectionLayoutSection(group: group)
        
        let layout = UICollectionViewCompositionalLayout(section: section)
        vehicleListCollectionView.collectionViewLayout = layout
    }
}
